src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java

```diff
@@ -0,0 +1,210 @@
+    private Handler<Integer> responseWatcher;
+    private final String attestationEndpoint;
+    private final HttpClient httpClient;
+    private final IClock clock;
+    private Vertx vertx;
+    private boolean isExpiryCheckScheduled;
+    private boolean isAttesting;
+    // Set this to be Instant.MAX so that if it's not set it won't trigger the re-attest
+    private Instant attestationTokenExpiresAt = Instant.MAX;
+    private final Lock lock;
+    private final AttestationTokenDecryptor attestationTokenDecryptor;
+
+    public AttestationTokenRetriever(String attestationEndpoint, ApplicationVersion appVersion, IAttestationProvider attestationProvider,
+                                     Handler<Integer> responseWatcher, IClock clock, HttpClient httpClient,
+                                     AttestationTokenDecryptor attestationTokenDecryptor) throws IOException {
+        this.attestationEndpoint = attestationEndpoint;
+        this.appVersion = appVersion;
+        this.attestationProvider = attestationProvider;
+        this.attestationToken = new AtomicReference<>(null);
+        this.responseWatcher = responseWatcher;
+        this.clock = clock;
+        this.lock = new ReentrantLock();
+        this.isExpiryCheckScheduled = false;
+        this.isAttesting = false;
+        if (httpClient == null) {
+            this.httpClient = HttpClient.newHttpClient();
+        } else {
+            this.httpClient = httpClient;
+        }
+        if (attestationTokenDecryptor == null) {
+            this.attestationTokenDecryptor = new AttestationTokenDecryptor();
+        } else {
+            this.attestationTokenDecryptor = attestationTokenDecryptor;
+        }
+    }
+
+    private void attestationExpirationCheck(long timerId) {
+        if (isAttesting) {
+            LOGGER.warn("In the process of attesting. Skip re-attest.");
+        } else {
+            Instant currentTime = clock.now();
+            Instant tenMinutesBeforeExpire = attestationTokenExpiresAt.minusSeconds(600);
+
+            if (currentTime.isAfter(tenMinutesBeforeExpire)) {
+                LOGGER.info("Attestation token is 10 mins from the expiry timestamp %s. Re-attest...", attestationTokenExpiresAt);
+                try {
+                    attest();
+                } catch (AttestationTokenRetrieverException | IOException e) {
+                    notifyResponseStatusWatcher(401);
+                    LOGGER.info("Re-attest failed: ", e.getMessage());
+                }
+            }
+        }
+    }
+
+    private void scheduleAttestationExpirationCheck() {
+        if (!this.isExpiryCheckScheduled) {
+            // Schedule the task to run every minute
+            this.vertx.setPeriodic(0, 60000, this::attestationExpirationCheck);
+            this.isExpiryCheckScheduled = true;
+        }
+    }
+
+    public void attest() throws IOException, AttestationTokenRetrieverException {
+        try {
+            this.isAttesting = true;
```

Comment:
I suggest we move this to inside the attestationExpirationCheck, and put it just before the call to attest (line 78), and then in a finally block, set it to false.
As it is here, if attesting throws an error, the flag will never get reset to false, so it will never re-attest